### PR TITLE
Flexible array pattern

### PR DESCRIPTION
**Pull request by chriseth**
Depends on #1205 

Will document in https://github.com/powdr-labs/powdr/pull/1214

### CODE REVIEW

File: pil-analyzer/src/expression_processor.rs [hunk 1]
```diff
@@ -142,8 +142,14 @@ impl<D: AnalysisDriver> ExpressionProcessor<D> {
     /// Processes a pattern, registering all variables bound in there.
     fn process_pattern(&mut self, pattern: &Pattern) {
         match pattern {
-            Pattern::CatchAll | Pattern::Number(_) | Pattern::String(_) => {}
+            Pattern::CatchAll | Pattern::Rest | Pattern::Number(_) | Pattern::String(_) => {}
             Pattern::Tuple(items) | Pattern::Array(items) => {
+                if matches!(pattern, Pattern::Array(_)) {
+                    // If there is more than one Pattern::Rest in items, it is an error
+                    if items.iter().filter(|p| *p == &Pattern::Rest).count() > 1 {
+                        panic!("Only one \"..\"-item allowed in array pattern");
```
Comment (pacheco) [inline pil-analyzer/src/expression_processor.rs:150]: maybe add a "should fail" test for this case?

File: ast/src/parsed/mod.rs [hunk 2]
```diff
@@ -824,7 +824,8 @@ impl Children<Expression> for ArrayExpression {
 
 #[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Serialize, Deserialize, JsonSchema)]
 pub enum Pattern {
-    CatchAll,
+    CatchAll, // "_", matches a single value
+    Rest,     // "..", matches a series of values, only valid inside array patterns
```
Comment (pacheco) [inline ast/src/parsed/mod.rs:828]: maybe call this pattern `Ellipsis` instead of `Rest`?

File: pil-analyzer/src/statement_processor.rs [hunk 3]
```diff
@@ -523,41 +505,13 @@ where
...
-            index: self
-                .evaluate_expression_to_int(index)
+            index: untyped_evaluator::evaluate_expression_to_int(self.driver, index)
                 .unwrap()
                 .try_into()
                 .unwrap(),
         })]
     }
 
-    /// Turns a Type<Expression> to a Type<u64> by evaluating the array length expressions.
```
Comment (chriseth) [inline pil-analyzer/src/statement_processor.rs:534]: I think this will go away once the basic patterns are merged and the merge target changes to "main".

**Review by pacheco — APPROVED** [review]
LGTM

**Pull request merged**